jupyterhub_traefik_proxy/redis.py

```diff
@@ -0,0 +1,166 @@
+"""Redis backend"""
+import asyncio
+from urllib.parse import urlparse
+
+from traitlets import Any, Dict, Unicode, default
+
+from .kv_proxy import TKvProxy
+from .traefik_utils import deep_merge
+
+
+class TraefikRedisProxy(TKvProxy):
+    """JupyterHub Proxy implementation using traefik and redis"""
+
+    provider_name = "redis"
+
+    redis_url = Unicode(
+        'redis://localhost:6379', config=True, help="""The URL for the redis endpoint"""
+    )
+    redis_username = Unicode(config=True, help="The redis username")
+    redis_password = Unicode(config=True, help="The redis password")
+
+    redis_client_kwargs = Dict(
+        config=True,
+        help="Additional keyword arguments to pass through to the `redis.Redis` constructor",
+    )
+
+    redis = Any()
+
+    @default("redis")
+    def _connect_redis(self):
+        try:
+            from redis.asyncio import Redis
+        except ImportError:
+            raise ImportError(
+                "Please install `redis` package to use traefik-proxy with redis"
+            )
+
+        url = urlparse(self.redis_url)
+        if url.port:
+            port = url.port
+        else:
+            # default port
+            port = 6379
+        kwargs = dict(
+            host=url.hostname,
+            port=port,
+            decode_responses=True,
+        )
+        if self.redis_password:
+            kwargs["password"] = self.redis_password
+        if self.redis_username:
+            kwargs["username"] = self.redis_username
+        kwargs.update(self.redis_client_kwargs)
+        return Redis(**kwargs)
+
+    async def _cleanup(self):
+        f = super()._cleanup()
+        if f is not None:
+            await f
+        await self.redis.close()
+
+    def _setup_traefik_static_config(self):
+        self.log.debug("Setting up the redis provider in the static config")
```

Comment:
```suggestion
        self.log.debug("Setting up the redis provider in the traefik static config")
```

Too many pieces of config, being explicit is helpful - like the function name is!